Python/Product/Analysis/Intellisense/LinePreservingCodeReplacer.cs

```diff
@@ -26,124 +27,89 @@ namespace Microsoft.PythonTools.Intellisense {
+                    } else if (diff.NewLength > length) {
+                        edits.Add(Insert(diff.OldStart + length, diff.NewStart + length, diff.NewEnd));
                     }
                 }
-
-                if (!replaced) {
-                    ReplaceLines(edits, curOldLine, _oldLines.Count, startNewLine, _newLines.Count);
-                    curOldLine = _oldLines.Count;
-                    break;
-                }
             }
 
-            if (curOldLine < _oldLines.Count) {
-                // remove the remaining new lines
-                edits.Add(
-                    DocumentChange.Delete(new SourceSpan(
-                        _oldLines[curOldLine].SourceStart,
-                        _oldLines[_oldLines.Count - 1].SourceEnd
-                    ))
-                );
-            }
-            return edits.ToArray();
+            return edits.AsReadOnly();
         }
 
-        private void ReplaceLines(List<DocumentChange> edits, int startOldLine, int endOldLine, int startNewLine, int endNewLine) {
-            int oldLineCount = endOldLine - startOldLine;
-            int newLineCount = endNewLine - startNewLine;
-
-            // replace one line at a time instead of all of the lines at once so that we preserve breakpoints
-            int excessNewLineStart = startNewLine - startOldLine;
-            for (int i = startOldLine; i < endOldLine && i < (endNewLine - startNewLine + startOldLine); i++) {
-                edits.Add(
-                    DocumentChange.Replace(
-                        _oldLines[i].SourceExtent,
-                        GetNewText(startNewLine + i - startOldLine)
-                    )
-                );
-                excessNewLineStart = startNewLine + i - startOldLine + 1;
-            }
+        private DocumentChange Replace(int oldLineIndex, int newLineIndex) {
+            var oldLineInfo = _oldLines[oldLineIndex].Info;
+            var newLineInfo = _newLines[newLineIndex].Info;
+            var newSource = _newLines[newLineIndex].Source;
+            var text = newSource.Substring(newLineInfo.Start, newLineInfo.Length);
+            return DocumentChange.Replace(oldLineInfo.SourceStart, oldLineInfo.SourceEnd, text);
+        }
+
+        private DocumentChange Insert(int oldStart, int newStart, int newEnd) {
+            var startIndex = _newLines[newStart].Info.Start;
+            var length = _newLines[newEnd].Info.EndIncludingLineBreak - startIndex;
+            var newSource = _newLines[newStart].Source;
+            var insertText = new StringBuilder(newSource, startIndex, length, length)
+                .Replace(NewLineKind.LineFeed.GetString(), _newLine)
+                .Replace(NewLineKind.CarriageReturn.GetString(), _newLine)
+                .Replace(NewLineKind.CarriageReturnLineFeed.GetString(), _newLine)
+                .ToString();
+
+            var insertLocation = _oldLines[oldStart].Info.SourceEndIncludingLineBreak;
```

Comment:
Don't know if it matters (probably not), but this location could be different from `_oldLines[oldStart + 1].Info.SourceStart` if any edits race. I don't think they can, but semantically you may prefer one over the other.

Reply:
It is actually should be `_oldLines[oldStart].Info.SourceStart` for all lines except the last one, which should be `_oldLines[oldStart-1].Info.SourceEndIncludingLineBreak`. Is there any formatting rule that can insert the line?